openhtf/util/validators.py

```diff
@@ -50,124 +50,140 @@ def MyPhase(test):
-      return isinstance(other, type(self)) and self.regex == other.regex
-
-    def __ne__(self, other):
-      return not self == other
-
-  # We have to use our saved reference to the re module because this module
-  # has lost all references by the sys.modules replacement and has been gc'd.
-  def matches_regex(self, regex):
-    return self.RegexMatcher(regex, self.modules['re'].compile(regex))
-
-sys.modules[__name__] = Validators(sys.modules)
+from openhtf import util
+
+_VALIDATORS = {}
+
+
+def register(validator, name=None):
+  name = name or validator.__name__
+  if name in _VALIDATORS:
+    raise ValueError('Duplicate validator name', name)
+  _VALIDATORS[name] = validator
+
+
+def create_validator(name, *args, **kwargs):
+  return _VALIDATORS[name](*args, **kwargs)
+
+# Built-in validators below this line
+
+class InRange(object):
+  """Validator to verify a numeric value is within a range."""
+
+  def __init__(self, minimum, maximum, type=None):
+    if minimum is None and maximum is None:
+      raise ValueError('Must specify minimum, maximum, or both')
+    if (minimum is not None and maximum is not None
+        and isinstance(minimum, numbers.Number)
+        and isinstance(maximum, numbers.Number)
+        and minimum > maximum):
+      raise ValueError('Minimum cannot be greater than maximum')
+    self.minimum = minimum
+    self.maximum = maximum
+    self._type = type or (lambda x: x)
+
+  def with_args(self, **kwargs):
+    return type(self)(
+        minimum=util.format_string(self.minimum, kwargs),
+        maximum=util.format_string(self.maximum, kwargs),
+        type=self._type,
+    )
+
+  def __call__(self, value):
+    if value is None:
+      return False
+    import math
+    # Check for nan
+    if math.isnan(value):
+      return False
+    minimum = self._type(self.minimum)
+    maximum = self._type(self.maximum)
+    # Check for equal bounds first so we can use with non-numeric values.
+    if minimum == maximum and value != minimum:
+      return False
+    if minimum is not None and value < minimum:
+      return False
+    if maximum is not None and value > maximum:
+      return False
+    return True
+
+  def __str__(self):
+    assert self.minimum is not None or self.maximum is not None
+    if self.minimum is not None and self.maximum is not None:
+      if self.minimum == self.maximum:
+        return 'x == %s' % self.minimum
+      return '%s <= x <= %s' % (self.minimum, self.maximum)
+    if self.minimum is not None:
+      return '%s <= x' % self.minimum
+    if self.maximum is not None:
+      return 'x <= %s' % self.maximum
+
+  def __eq__(self, other):
+    return (isinstance(other, type(self)) and
+            self.minimum == other.minimum and self.maximum == other.maximum)
+
+  def __ne__(self, other):
+    return not self == other
+
+in_range = InRange  # pylint: disable=invalid-name
+register(in_range, name='in_range')
+
+
+@register
+def equals(self, value, type=None):
```

Comment:
It's to allow users to do this:

```py
@htf.measures(htf.Measurement('xyz').in_range('{voltage}', 20, type=float))
def Something(...):
  ...

Something.with_args(voltage=12)
```

That way, `Something`'s measurement 'xyz' will have a validator based on the arguments, but with the right type. We can't assume a particular type in case of precision errors